Duplicati/Library/Snapshots/WindowsSnapshot.cs

```diff
@@ -108,59 +105,94 @@ public WindowsSnapshot(string[] sourcepaths, Dictionary<string, string> options)
-                    if (!Utility.Utility.ParseBoolOption(options, "hyperv-backup-no-merge"))
-                        hyperVUtility.MergeVhd(requestedHyperVMs);
+                    var hyperVUtility = new HyperVUtility();
+                    var hyperVGuests = hyperVUtility.GetHyperVGuests();
 
-                    //Option: Create-new-vm-after-restore, 
-                    //hyperVUtility.CreateHyperVMachine("TestVM", xmlpath);
-                    #endregion Testing Features
-                }
+                    Logging.Log.WriteMessage(string.Format("Found {0} virtual machines on Hyper-V.", hyperVGuests.Count), Logging.LogMessageType.Information);
 
-                m_backup.StartSnapshotSet();
+                    bool bNotFound = false;
+                    List<string> notFoundVM = new List<string>();
+                    foreach (var requestedHyperVM in requestedHyperVMs)
+                    {
+                        var foundVMs = hyperVGuests.FindAll(x => (x.ID.ToLower() == requestedHyperVM.ToLower()));
 
-                try
-                {
-                    //Gather information on all Vss writers
-                    m_backup.GatherWriterMetadata();
+                        if (foundVMs.Count != 1)
+                        {
+                            bNotFound = true;
+                            notFoundVM.Add(requestedHyperVM);
+                            Logging.Log.WriteMessage(string.Format("Cannot find virtual machine with ID {0} on Hyper-V.", requestedHyperVM), Logging.LogMessageType.Error);
+                        }
+                        else
+                            Logging.Log.WriteMessage(string.Format("Found virtual machine {0} with ID {1} on Hyper-V.", foundVMs[0].Name, requestedHyperVM), Logging.LogMessageType.Information);
+                    }
 
-                    //Update the sourcepaths if we're backing up Hyper-V Machines
-                    if (requestedHyperVMs.Count > 0)
+                    if (bNotFound)
+                        throw new Exception(string.Format("Cannot find virtual machine with ID {0} on Hyper-V.", string.Join(", ", notFoundVM.ToArray())));
+
+                    var productType = new ManagementObjectSearcher("SELECT * FROM Win32_OperatingSystem")
+                        .Get().OfType<ManagementObject>()
+                        .Select(o => (uint)o.GetPropertyValue("ProductType")).First();
+
+                    if (productType != 1) // Hyper-V writer is present only on Server version of Windows
                     {
-                        //Find sourcepaths in Hyper-V WriterMetaData
-                        var m_backup_wmd = m_backup.WriterMetadata.FirstOrDefault(o => o.WriterId.Equals(HyperVWriterGuid));
-                        if (m_backup_wmd != null)
-                            foreach (var component in m_backup_wmd.Components)
+                        try
+                        {
+                            m_backup.GatherWriterMetadata();

```

Comment:
It looks like this call is now only made if `productType != 1`, where it was previously executed always.
I think this should be called regardless of the existence of Hyper-V?